README.md
[![Latest Stable Version](http://poser.pugx.org/nguyenanhung/codeigniter3-swoole/v)](https://packagist.org/packages/nguyenanhung/codeigniter3-swoole) [![Total Downloads](http://poser.pugx.org/nguyenanhung/codeigniter3-swoole/downloads)](https://packagist.org/packages/nguyenanhung/codeigniter3-swoole) [![Latest Unstable Version](http://poser.pugx.org/nguyenanhung/codeigniter3-swoole/v/unstable)](https://packagist.org/packages/nguyenanhung/codeigniter3-swoole) [![License](http://poser.pugx.org/nguyenanhung/codeigniter3-swoole/license)](https://packagist.org/packages/nguyenanhung/codeigniter3-swoole) [![PHP Version Require](http://poser.pugx.org/nguyenanhung/codeigniter3-swoole/require/php)](https://packagist.org/packages/nguyenanhung/codeigniter3-swoole)

# Swoole Adapter for CodeIgniter v3 Framework

You want long-run task? timers? FPM to CLI? Code reusing in both FPM & CLI mode?

"It's so easy!"

This adapter would make it so easy to using swoole within Codeigniter framework.

With this adapter, you can start a task(CLI) anywhere (FPM) you want from your code.

That's means you can start a CLI task from a FPM process.

## Install

```shell
composer require nguyenanhung/codeigniter3-swoole
```

## How to

1. first, of course you must install `codeigniter3-swoole` to your codeigniter project.
2. (this step is option) copy these two config files `swoole.php` and `timers.php` from `config` to your `app/config` folder.
3. start swoole server `php index.php swoole/server/start`
4. you can use `\nguyenanhung\CodeIgniter\Swoole\Core\Client::send($data)` to start a task now!
5. there's no step 5.

## What is a task?

A task is just a method of your codeigniter controlloer, so almost any controller method can be used as a task.

Let's see the code

```php
\nguyenanhung\CodeIgniter\Swoole\Core\Client::send(
[
    'route'  => 'your/route/uri/to/a/method'
    'params' => ['test' => 666]
]);
```

The `route` is used for find which method to be call as a task, and `params` is the parameters array that you may want to pass to the task.

So, that's all of it!

## Server CLI Commands

```shell

// start the swoole server
php index.php swoole/server/start

// stop the swoole server
php index.php swoole/server/stop

// reload all wokers of swoole server
php index.php swoole/server/reload

```

## A little more

The step 2 copied files were config files for this adapter.

`swoole.php` file can set host, port, log file and so on.

`timers.php` file can set some timer methods for swoole server, these timers will be started once the server inited.

The demos are same as below shows.

```php
class Test extends CI_Controller
{

    // ------------------------------------------------------------------------------

    /**
     * here's the task 'tests/test/task'
     */
    public function task()
    {
        $data = $this->input->post();     // as you see, params worked like normally post data

        log_message('info', var_export($data, true));
    }

    // ------------------------------------------------------------------------------

    /**
     * here's the timer method
     *
     * you should copay timers.php to your config folder,
     * then add $timers['tests/test/task_timer'] = 10000; and start the swoole server.
     *
     * this method would be called every 10 seconds per time.
     */
    public function task_timer()
    {
        log_message('info', 'timer works!');
    }

    // ------------------------------------------------------------------------------

    /**
     * send data to task
     */
    public function send()
    {
        try
        {
            \nguyenanhung\CodeIgniter\Swoole\Core\Client::send(
            [
                'route'  => 'tests/test/task',
                'params' => ['hope' => 'it works!'],
            ]);
        }
        catch (\Exception $e)
        {
            log_message('error', $e->getMessage());
            log_message('error', $e->getTraceAsString());
        }
    }

    // ------------------------------------------------------------------------------

}
```

## Contact & Support

If any question & request, please contact following information

| Name        | Email                | Skype            | Facebook      |
|-------------|----------------------|------------------|---------------|
| Hung Nguyen | dev@nguyenanhung.com | nguyenanhung5891 | @nguyenanhung |

From Vietnam with Love <3


CHANGELOG.md
# CHANGELOG

## [1.0.0] - 2022/07/25

### Added

- [x] Init Project

## Contact & Support

If any question & request, please contact following information

| Name        | Email                | Skype            | Facebook      |
|-------------|----------------------|------------------|---------------|
| Hung Nguyen | dev@nguyenanhung.com | nguyenanhung5891 | @nguyenanhung |


test/TestSwoole.php
<?php defined('BASEPATH') OR exit('No direct script access allowed');

/**
 * Class TestSwoole
 *
 * @author    713uk13m <dev@nguyenanhung.com>
 * @copyright 713uk13m <dev@nguyenanhung.com>
 */
class TestSwoole extends CI_Controller
{

    // ------------------------------------------------------------------------------

    /**
     * here's the task
     */
    public function task()
    {
        $data = $this->input->post();

        log_message('error', var_export($data, true));
    }

    // ------------------------------------------------------------------------------

    /**
     * here's the timer method
     */
    public function task_timer()
    {
        log_message('error', 'timer works!');
    }

    // ------------------------------------------------------------------------------

    /**
     * send data to task
     */
    public function send()
    {
        try
        {
            \CiSwoole\Core\Client::send(
                [
                    'route'  => 'testswoole/task',
                    'params' => ['hope' => 'it works!'],
                ]);
        }
        catch (\Exception $e)
        {
            log_message('error', $e->getMessage());
            log_message('error', $e->getTraceAsString());
        }
    }

    // ------------------------------------------------------------------------------

}


config/swoole.php
<?php
defined('BASEPATH') or exit('No direct script access allowed');

/*
|--------------------------------------------------------------------------
| Swoole Server Host
|--------------------------------------------------------------------------
|
| For custom configuration, you can copy this swoole.php to
| the application/config folder.
|
| Example for tcp:
| $swoole['server_type'] = SWOOLE_SOCK_TCP;
| $swoole['server_host'] = '127.0.0.1';
| $swoole['server_port'] = 9501;
|
| Warning:
|    Do not change any configuration, unless you know what it is!
|
| For more detail document, view this link below
| @link https://openswoole.com/docs/modules/swoole-server-doc
| @link https://wiki.swoole.com/wiki/page/476.html
*/
$swoole['server_type'] = SWOOLE_SOCK_UNIX_STREAM;
$swoole['server_host'] = '/var/run/swoole.sock';
$swoole['server_port'] = null;


/*
|--------------------------------------------------------------------------
| Connection
|--------------------------------------------------------------------------
|
| Max connection number
*/
$swoole['max_conn'] = 1000;


/*
|--------------------------------------------------------------------------
| Request
|--------------------------------------------------------------------------
|
| Exit worker when finished ? times
*/
$swoole['max_request'] = 10;


/*
|--------------------------------------------------------------------------
| Waite Time
|--------------------------------------------------------------------------
|
| After ? seconds force exit
*/
$swoole['max_wait_time'] = 60;


/*
|--------------------------------------------------------------------------
| Worker
|--------------------------------------------------------------------------
|
| How many workers did you want?
*/
$swoole['worker_num'] = 4;


/*
|--------------------------------------------------------------------------
| Swoole Server Port
|--------------------------------------------------------------------------
|
| 3: post to a free worker
*/
$swoole['dispatch_mode'] = 3;


/*
|--------------------------------------------------------------------------
| Task Worker
|--------------------------------------------------------------------------
|
| Worker numbers for task
*/
$swoole['task_worker_num'] = 12;


/*
|--------------------------------------------------------------------------
| Logs
|--------------------------------------------------------------------------
|
| Server log file & debug file
*/
$swoole['log_file']   = APPPATH . 'logs/swoole.log';
$swoole['debug_file'] = APPPATH . 'logs/swoole_debug.log';


/*
|--------------------------------------------------------------------------
| Swoole Process
|--------------------------------------------------------------------------
|
| Process name for swoole
*/
$swoole['process_name'] = 'CodeIgniter3Swoole';


/*
|--------------------------------------------------------------------------
| Heartbeat
|--------------------------------------------------------------------------
|
| Seconds for heartbeat
*/
$swoole['heartbeat_idle_time']      = 10;
$swoole['heartbeat_check_interval'] = 5;


// the end
return $swoole;


config/starter.php
<?php
/**
 * Codeigniter
 *
 * @link https://github.com/bcit-ci/CodeIgniter/blob/develop/system/core/CodeIgniter.php
 * @since Version 3.2.0
 * @change 2018/09/25
 */


/*
 * ------------------------------------------------------
 *  Start the timer... tick tock tick tock...
 * ------------------------------------------------------
 */
$BM =& load_class('Benchmark', 'core');
$BM->mark('total_execution_time_start');
$BM->mark('loading_time:_base_classes_start');
/*
 * ------------------------------------------------------
 *  Instantiate the config class
 * ------------------------------------------------------
 *
 * Note: It is important that Config is loaded first as
 * most other classes depend on it either directly or by
 * depending on another class that uses it.
 *
 */
$CFG =& load_class('Config', 'core');
// Do we have any manually set config items in the index.php file?
if (isset($assign_to_config) && is_array($assign_to_config))
{
    foreach ($assign_to_config as $key => $value)
    {
        $CFG->set_item($key, $value);
    }
}
/*
 * ------------------------------------------------------
 *  Instantiate the hooks class
 * ------------------------------------------------------
 */
$EXT =& load_class('Hooks', 'core', $CFG);
/*
 * ------------------------------------------------------
 *  Is there a "pre_system" hook?
 * ------------------------------------------------------
 */
$EXT->call_hook('pre_system');
/*
 * ------------------------------------------------------
 * Important charset-related stuff
 * ------------------------------------------------------
 *
 * Configure mbstring and/or iconv if they are enabled
 * and set MB_ENABLED and ICONV_ENABLED constants, so
 * that we don't repeatedly do extension_loaded() or
 * function_exists() calls.
 *
 * Note: UTF-8 class depends on this. It used to be done
 * in it's constructor, but it's _not_ class-specific.
 *
 */
$charset = strtoupper(config_item('charset'));
ini_set('default_charset', $charset);
if (extension_loaded('mbstring'))
{
    !defined('MB_ENABLED') AND define('MB_ENABLED', TRUE);
    // mbstring.internal_encoding is deprecated starting with PHP 5.6
    // and it's usage triggers E_DEPRECATED messages.
    @ini_set('mbstring.internal_encoding', $charset);
    // This is required for mb_convert_encoding() to strip invalid characters.
    // That's utilized by CI_Utf8, but it's also done for consistency with iconv.
    mb_substitute_character('none');
}
else
{
    !defined('MB_ENABLED') AND define('MB_ENABLED', FALSE);
}
// There's an ICONV_IMPL constant, but the PHP manual says that using
// iconv's predefined constants is "strongly discouraged".
if (extension_loaded('iconv'))
{
    !defined('ICONV_ENABLED') AND define('ICONV_ENABLED', TRUE);
    // iconv.internal_encoding is deprecated starting with PHP 5.6
    // and it's usage triggers E_DEPRECATED messages.
    @ini_set('iconv.internal_encoding', $charset);
}
else
{
    !defined('ICONV_ENABLED') AND define('ICONV_ENABLED', FALSE);
}
if (is_php('5.6'))
{
    ini_set('php.internal_encoding', $charset);
}
/*
 * ------------------------------------------------------
 *  Load compatibility features
 * ------------------------------------------------------
 */
require_once(BASEPATH.'core/compat/mbstring.php');
require_once(BASEPATH.'core/compat/hash.php');
require_once(BASEPATH.'core/compat/password.php');
require_once(BASEPATH.'core/compat/standard.php');
/*
 * ------------------------------------------------------
 *  Instantiate the UTF-8 class
 * ------------------------------------------------------
 */
$UNI =& load_class('Utf8', 'core', $charset);
/*
 * ------------------------------------------------------
 *  Instantiate the URI class
 * ------------------------------------------------------
 */
$URI =& load_class('URI', 'core', $CFG);
/*
 * ------------------------------------------------------
 *  Instantiate the routing class and set the routing
 * ------------------------------------------------------
 */
$RTR =& load_class('Router', 'core', isset($routing) ? $routing : NULL);
/*
 * ------------------------------------------------------
 *  Instantiate the output class
 * ------------------------------------------------------
 */
$OUT =& load_class('Output', 'core');
/*
 * ------------------------------------------------------
 *	Is there a valid cache file? If so, we're done...
 * ------------------------------------------------------
 */
if ($EXT->call_hook('cache_override') === FALSE && $OUT->_display_cache($CFG, $URI) === TRUE)
{
    exit;
}
/*
 * -----------------------------------------------------
 * Load the security class for xss and csrf support
 * -----------------------------------------------------
 */
$SEC =& load_class('Security', 'core', $charset);
/*
 * ------------------------------------------------------
 *  Load the Input class and sanitize globals
 * ------------------------------------------------------
 */
$IN =& load_class('Input', 'core', $SEC);
/*
 * ------------------------------------------------------
 *  Load the Language class
 * ------------------------------------------------------
 */
$LANG =& load_class('Lang', 'core');
/*
 * ------------------------------------------------------
 *  Load the app controller and local controller
 * ------------------------------------------------------
 *
 */
// Load the base controller class
require_once BASEPATH.'core/Controller.php';
/**
 * Reference to the CI_Controller method.
 *
 * Returns current CI instance object
 *
 * @return CI_Controller
 */
if (! function_exists('get_instance'))   // check
{
    function &get_instance()
    {
        return CI_Controller::get_instance();
    }
}
if (file_exists(APPPATH.'core/'.$CFG->config['subclass_prefix'].'Controller.php'))
{
    require_once APPPATH.'core/'.$CFG->config['subclass_prefix'].'Controller.php';
}
// Set a mark point for benchmarking
$BM->mark('loading_time:_base_classes_end');
/*
 * ------------------------------------------------------
 *  Sanity checks
 * ------------------------------------------------------
 *
 *  The Router class has already validated the request,
 *  leaving us with 3 options here:
 *
 *	1) an empty class name, if we reached the default
 *	   controller, but it didn't exist;
 *	2) a query string which doesn't go through a
 *	   file_exists() check
 *	3) a regular request for a non-existing page
 *
 *  We handle all of these as a 404 error.
 *
 *  Furthermore, none of the methods in the app controller
 *  or the loader class can be called via the URI, nor can
 *  controller methods that begin with an underscore.
 */
$e404 = FALSE;
$class = ucfirst($RTR->class);
$method = $RTR->method;
if (empty($class) OR ! file_exists(APPPATH.'controllers/'.$RTR->directory.$class.'.php'))
{
    $e404 = TRUE;
}
else
{
    require_once(APPPATH.'controllers/'.$RTR->directory.$class.'.php');
    if ( ! class_exists($class, FALSE) OR $method[0] === '_' OR method_exists('CI_Controller', $method))
    {
        $e404 = TRUE;
    }
    elseif (method_exists($class, '_remap'))
    {
        $params = array($method, array_slice($URI->rsegments, 2));
        $method = '_remap';
    }
    elseif ( ! method_exists($class, $method))
    {
        $e404 = TRUE;
    }
    /**
     * DO NOT CHANGE THIS, NOTHING ELSE WORKS!
     *
     * - method_exists() returns true for non-public methods, which passes the previous elseif
     * - is_callable() returns false for PHP 4-style constructors, even if there's a __construct()
     * - method_exists($class, '__construct') won't work because CI_Controller::__construct() is inherited
     * - People will only complain if this doesn't work, even though it is documented that it shouldn't.
     *
     * ReflectionMethod::isConstructor() is the ONLY reliable check,
     * knowing which method will be executed as a constructor.
     */
    elseif ( ! is_callable(array($class, $method)))
    {
        $reflection = new ReflectionMethod($class, $method);
        if ( ! $reflection->isPublic() OR $reflection->isConstructor())
        {
            $e404 = TRUE;
        }
    }
}
if ($e404)
{
    if ( ! empty($RTR->routes['404_override']))
    {
        if (sscanf($RTR->routes['404_override'], '%[^/]/%s', $error_class, $error_method) !== 2)
        {
            $error_method = 'index';
        }
        $error_class = ucfirst($error_class);
        if ( ! class_exists($error_class, FALSE))
        {
            if (file_exists(APPPATH.'controllers/'.$RTR->directory.$error_class.'.php'))
            {
                require_once(APPPATH.'controllers/'.$RTR->directory.$error_class.'.php');
                $e404 = ! class_exists($error_class, FALSE);
            }
            // Were we in a directory? If so, check for a global override
            elseif ( ! empty($RTR->directory) && file_exists(APPPATH.'controllers/'.$error_class.'.php'))
            {
                require_once(APPPATH.'controllers/'.$error_class.'.php');
                if (($e404 = ! class_exists($error_class, FALSE)) === FALSE)
                {
                    $RTR->directory = '';
                }
            }
        }
        else
        {
            $e404 = FALSE;
        }
    }
    // Did we reset the $e404 flag? If so, set the rsegments, starting from index 1
    if ( ! $e404)
    {
        $class = $error_class;
        $method = $error_method;
        $URI->rsegments = array(
            1 => $class,
            2 => $method
        );
    }
    else
    {
        show_404($RTR->directory.$class.'/'.$method);
    }
}
if ($method !== '_remap')
{
    $params = array_slice($URI->rsegments, 2);
}
/*
 * ------------------------------------------------------
 *  Is there a "pre_controller" hook?
 * ------------------------------------------------------
 */
$EXT->call_hook('pre_controller');
/*
 * ------------------------------------------------------
 *  Instantiate the requested controller
 * ------------------------------------------------------
 */
// Mark a start point so we can benchmark the controller
$BM->mark('controller_execution_time_( '.$class.' / '.$method.' )_start');
$CI = new $class();
/*
 * ------------------------------------------------------
 *  Is there a "post_controller_constructor" hook?
 * ------------------------------------------------------
 */
$EXT->call_hook('post_controller_constructor');
/*
 * ------------------------------------------------------
 *  Call the requested method
 * ------------------------------------------------------
 */
call_user_func_array(array(&$CI, $method), $params);
// Mark a benchmark end point
$BM->mark('controller_execution_time_( '.$class.' / '.$method.' )_end');
/*
 * ------------------------------------------------------
 *  Is there a "post_controller" hook?
 * ------------------------------------------------------
 */
$EXT->call_hook('post_controller');
/*
 * ------------------------------------------------------
 *  Send the final rendered output to the browser
 * ------------------------------------------------------
 */
if ($EXT->call_hook('display_override') === FALSE)
{
    $OUT->_display();
}
/*
 * ------------------------------------------------------
 *  Is there a "post_system" hook?
 * ------------------------------------------------------
 */
$EXT->call_hook('post_system');


/*
 * ------------------------------------------------------
 *  Finally Return An Array (required)
 * ------------------------------------------------------
 */
return [];


helpers/helpers.php
<?php
// ------------------------------------------------------------------------------
if (!function_exists('getCiSwooleConfig')) {
    /**
     * Function getCiSwooleConfig
     *
     * @param string $name
     *
     * @return array
     * @throws \Exception
     * @author   : 713uk13m <dev@nguyenanhung.com>
     * @copyright: 713uk13m <dev@nguyenanhung.com>
     * @time     : 25/07/2022 19:45
     */
    function getCiSwooleConfig(string $name)
    {
        $pathA = APPPATH . "config/{$name}.php";

        $pathB = dirname(__FILE__) . "/{$name}.php";

        $data = file_exists($pathA) ? include($pathA) : include($pathB);

        if (!is_array($data)) {
            throw new \Exception("{$name} config file must return as array");
        }

        return $data;
    }
}


// ------------------------------------------------------------------------------
if (!function_exists('____intercepter____')) {
    /**
     * CI swoole intercepter
     *
     */
    function ____intercepter____()
    {
        try {
            if (!is_cli()) {
                return;
            }

            $command = trim($_SERVER['argv'][1] ?? '', '/');

            switch ($command) {
                // start server
                case 'swoole/server/start':
                    $start = \nguyenanhung\CodeIgniter\Swoole\Core\Server::start();
                    die("Start Server: {$start}\n");

                // stop server
                case 'swoole/server/stop':
                    \nguyenanhung\CodeIgniter\Swoole\Core\Client::shutdown();
                    die("Stop Server.\n");

                // reload server
                case 'swoole/server/reload':
                    \nguyenanhung\CodeIgniter\Swoole\Core\Client::reload();
                    die("Reload Workers.\n");

                default:
                    return;
            }
        } catch (\Throwable $e) {
            $log = "{$e->getMessage()}\n";
            $log .= "{$e->getTraceAsString()}\n";

            die("Operation Failed.\n{$log}");
        }
    }

    // ------------------------------------------------------------------------------

    /**
     * Let's start perform magic tricks
     */
    ____intercepter____();
}


src/Core/Server.php
<?php
/**
 * Project codeigniter3-swoole
 * Created by PhpStorm
 * User: 713uk13m <dev@nguyenanhung.com>
 * Copyright: 713uk13m <dev@nguyenanhung.com>
 * Date: 25/07/2022
 * Time: 14:17
 */

namespace nguyenanhung\CodeIgniter\Swoole\Core;

class Server
{

    // ------------------------------------------------------------------------------

    /**
     * host config
     *
     * @var array
     */
    private static $cfgs =
        [
            'server_port' => null,
            'server_host' => '/var/run/swoole.sock',
            'server_type' => SWOOLE_SOCK_UNIX_STREAM,
            'debug_file'  => APPPATH . 'logs/swoole_debug.log',
        ];

    // ------------------------------------------------------------------------------

    /**
     * server config
     *
     * warning: do not change this
     *
     * @var array
     */
    private static $config =
        [
            'daemonize'      => true,        // using as daemonize?
            'package_eof'    => '☯',         // \u262F
            'reload_async'   => true,
            'open_eof_split' => true,
            'open_eof_check' => true,
        ];

    // ------------------------------------------------------------------------------

    /**
     * start a swoole server in cli
     *
     * @return mixed
     */
    public static function start()
    {
        self::initConfig();

        $serv = new \Swoole\Server
        (
            self::$cfgs['server_host'],
            self::$cfgs['server_port'],
            SWOOLE_PROCESS,
            self::$cfgs['server_type']
        );

        // init config
        $serv->set(self::$config);

        // listen on server init
        $serv->on('ManagerStart', [Server::class, 'onManagerStart']);
        $serv->on('WorkerStart', [Server::class, 'onWorkerStart']);
        $serv->on('Start', [Server::class, 'onMasterStart']);

        // listen on base event
        $serv->on('Connect', [Server::class, 'onConnect']);
        $serv->on('Receive', [Server::class, 'onReceive']);
        $serv->on('Finish', [Server::class, 'onFinish']);
        $serv->on('Close', [Server::class, 'onClose']);
        $serv->on('Task', [Server::class, 'onTask']);

        // start server
        return $serv->start();
    }

    // ------------------------------------------------------------------------------

    /**
     * listen on master start
     *
     * @param \Swoole\Server $serv
     */
    public static function onMasterStart(\Swoole\Server $serv)
    {
        if (self::$cfgs['server_port'] === null) {
            @chmod(self::$cfgs['server_host'], 0777);
        }

        @swoole_set_process_name($serv->setting['process_name'] . '-MASTER');

        $msg = "SWOOLE MASTER: {$serv->manager_pid}\n";

        error_log($msg, 3, self::$cfgs['debug_file']);
    }

    // ------------------------------------------------------------------------------

    /**
     * listen on manager start
     *
     * @param \Swoole\Server $serv
     */
    public static function onManagerStart(\Swoole\Server $serv)
    {
        @swoole_set_process_name($serv->setting['process_name'] . '-MANAGER');

        $msg = "SWOOLE MANAGER: {$serv->manager_pid}\n";

        error_log($msg, 3, self::$cfgs['debug_file']);
    }

    // ------------------------------------------------------------------------------

    /**
     * listen on workers start & set timers
     *
     * @param \Swoole\Server $serv
     * @param int            $workerId
     */
    public static function onWorkerStart(\Swoole\Server $serv, $workerId)
    {
        // set process name
        ($workerId >= $serv->setting['worker_num']) ?
            @swoole_set_process_name($serv->setting['process_name'] . '-TASK') :
            @swoole_set_process_name($serv->setting['process_name'] . '-WORKER');

        // when task start, return
        if ($serv->taskworker) {
            return;
        }

        // init all timers
        self::initTimers($serv);
    }

    // ------------------------------------------------------------------------------

    /**
     * listen on receive data
     *
     * @param \Swoole\Server $serv
     * @param int            $fd
     * @param int            $reactorId
     * @param string         $data
     */
    public static function onReceive(\Swoole\Server $serv, $fd, $reactorId, $data = '')
    {
        // close client
        $serv->close($fd);

        // format passed
        $data = str_replace(self::$config['package_eof'], '', $data);
        $data = unserialize($data);

        if (!$data) {
            return;
        }

        // check is command
        if (!empty($data['shutdown'])) {
            $serv->shutdown();

            return;
        }

        // reload command
        if (!empty($data['reload'])) {
            $serv->reload();

            return;
        }

        // start a task
        $serv->task($data);
    }

    // ------------------------------------------------------------------------------

    /**
     * listen on task
     *
     * @param \Swoole\Server $serv
     * @param int            $taskId
     * @param int            $workerId
     * @param array          $data
     */
    public static function onTask(\Swoole\Server $serv, $taskId, $workerId, $data)
    {
        try {
            $_SERVER['argv'] =
                [
                    0 => SELF,
                    1 => $data['route'],
                ];

            $_POST = $data['params'] ?? [];

            getCiSwooleConfig('starter');
        } // kill process
        catch (\Throwable $e) {
            self::logs($e);
        } finally {
            \Swoole\Process::kill(getmypid());
        }
    }

    // ------------------------------------------------------------------------------

    /**
     * listen on connect
     *
     * @param \Swoole\Server $serv
     * @param int            $fd
     */
    public static function onConnect(\Swoole\Server $serv, $fd)
    {
        return;
    }

    // ------------------------------------------------------------------------------

    /**
     * listen on close
     *
     * @param \Swoole\Server $serv
     * @param int            $fd
     */
    public static function onClose(\Swoole\Server $serv, $fd)
    {
        return;
    }

    // ------------------------------------------------------------------------------

    /**
     * listen on task finish
     *
     * @param \Swoole\Server $serv
     * @param int            $taskId
     * @param mixed          $data
     */
    public static function onFinish(\Swoole\Server $serv, $taskId, $data)
    {
        return;
    }

    // ------------------------------------------------------------------------------

    /**
     * init config
     *
     * @throws \Exception
     */
    private static function initConfig()
    {
        $config = getCiSwooleConfig('swoole');

        self::$cfgs['debug_file']  = $config['debug_file'];
        self::$cfgs['server_host'] = $config['server_host'];
        self::$cfgs['server_port'] = $config['server_port'];
        self::$cfgs['server_type'] = $config['server_type'];

        unset(
            $config['debug_file'], $config['server_host'],
            $config['server_port'], $config['server_type'],
        );

        self::$config = array_merge($config, self::$config);
    }

    // ------------------------------------------------------------------------------

    /**
     * log message to debug
     *
     * @param \Throwable $msg
     */
    private static function logs(\Throwable $msg)
    {
        $strings = $msg->getMessage() . "\n";
        $strings .= $msg->getTraceAsString();

        $time_nw = date('Y-m-d H:i:s');
        $content = "\n== {$time_nw} ============================\n";
        $content .= "{$strings}";
        $content .= "\n===================================================\n\n";

        error_log($content, 3, self::$cfgs['debug_file']);
    }

    // ------------------------------------------------------------------------------

    /**
     * init timers for stamsel
     *
     * @param \Swoole\Server $serv
     */
    private static function initTimers(\Swoole\Server $serv)
    {
        try {
            $timers = getCiSwooleConfig('timers');

            foreach ($timers as $route => $microSeconds) {
                $data =
                    [
                        'route'  => $route,
                        'params' => [],
                    ];

                $serv->tick($microSeconds, function () use ($serv, $data) {
                    $stats = $serv->stats();

                    if ($stats['tasking_num'] < 4) {
                        $serv->task($data);
                    }
                });
            }
        } catch (\Throwable $e) {
            self::logs($e);
        } finally {
            unset($timers);
        }
    }

    // ------------------------------------------------------------------------------

}
